This is to build the script for checking water quality in visualization.py.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

plt.style.use(['seaborn-colorblind', 'seaborn-talk'])
plt.rcParams['figure.dpi'] = 200
plt.rcParams['savefig.dpi'] = 200
plt.rcParams['savefig.bbox'] = 'tight'
plt.rcParams['font.sans-serif'] = ['Taipei Sans TC Beta'] # to plot Chinese words properly

import datetime
date = datetime.datetime.now().strftime('%Y%m%d')

In [4]:
wa_dir = 'data/database_ZAF_wa_merged_20211031.xlsx'
excel_dir = 'data/stds_and_cols.xlsx'
output_dir = 'results/',
excel_cols = ['項目', '單位', '飲用水水源水質標準第五條',
'飲用水水源水質標準第六條', '地下水污染監測標準第一類', 
'地下水污染監測標準第二類', '地下水污染管制標準第一類', 
'地下水污染管制標準第二類', '灌溉用水水質標準', 
'再生水用於工業用途水質基礎建議值一', 
'再生水用於工業用途水質基礎建議值二']

wa_df = pd.read_excel(wa_dir, parse_dates=['日期時間'])
wa_df['日期'] = pd.to_datetime([_.strftime('%Y-%m-%d') for _ in wa_df['日期時間']])
wa_df['井號'] = wa_df['井號'].astype(str)
std_df = pd.read_excel(excel_dir, usecols=excel_cols)

In [ ]:
import visualization

# use the 2021-05-01 to 2021-05-15 as the toy data
df = pd.read_csv('data/test.csv')
select = visualization.Waterlvl()
df = select.MarkbyEP(df=df)
std_name = '再生水用於工業用途水質基礎建議值一'

# Bugs fixing
1. Wrong asignment?<br>
`site_name = X['井名'][0]` trigers the error that no idex is 0.<br>
Solution: add `.reset_index(drop=True)` behind `X = wa_df[wa_df['井號'] == siteid]`.

2. Empty slicing dataframe?<br>
After `X = wa_df[wa_df['井號'] == siteid].reset_index(drop=True)`, the dataframe `X` has nothing, which causes the later codes error.<br>
Solution: check the siteid. They really don't have quality measurement, so I skip them but still print them out in case they need double chckeck.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 148 entries, 2021-05-14 23:50:00 to 2021-05-06 12:40:00
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   井名        148 non-null    object
 1   井號        148 non-null    object
 2   水面至井口深度   148 non-null    object
 3   Lon       148 non-null    object
 4   Lat       148 non-null    object
 5   月         148 non-null    object
 6   wl_check  148 non-null    bool  
dtypes: bool(1), object(6)
memory usage: 8.2+ KB


In [11]:
check = True
check_list=[]
out_df = pd.DataFrame()
for siteid in df['井號'].astype(str).unique():
    # collect the analytes having value in the standard (std_name) and '日期'
    cols = np.hstack([std_df.loc[~std_df[std_name].isna(), '項目'], '日期'])
    # select the data points of that siteid
    X = wa_df[wa_df['井號'] == siteid].reset_index(drop=True)
    # pick up the site name
    print(siteid)
    site_name = X['井名'][0]

7010111
10050111
10070111
9160212
7230312
9190142
7050122
9130122
9180121
7100211
9020111
7040121
7220231
9010212
7020111
9180111
7220221
9130222
10090121
9170421
9070311
9170411
7100221
9130211
10090111
9050121
9010112
7010211
9010122
7240122
9110211
9050111
9200111
7050112
9180311
7080132
7240112
9120121
9180221
7080141
7230332
9120131
7060121
9190112
9080111
8040222
9180131
7100112
7230322
9130112
7220211
7080131
9130232
7210121
10050121
7020121
9170122
7140121
7200132
9150131
9170112
7140131
10090131
9160141
9140122
8060211
7040141
9200221
9190132
7050132
9160222
9070112
8060121
7200112
10050131
7200121
9060212


KeyError: 0

In [12]:
X

Empty DataFrame
Columns: [井名, 資料來源, 日期時間, 井號, 緯度, 經度, 溫度, 電導度, 氨氮, 氫離子濃度指數, 錳, 鐵, 氯鹽, 硫酸鹽, 總有機碳, 鋅, 砷, 銅, 鎘, 鉻, 鈣, 鉛, 鈉, 鎂, 鉀, 總硬度, 總溶解固體物, 亞硝酸鹽氮, 井別, 鹼度, 水面至井口深度, 氧化還原電位, 硝酸鹽氮, 大腸桿菌群, 硫化物, 碳酸氫根, 鎳, 汞, 酚類, 氟鹽, 碳酸根, 溶氧量, 甲苯, 氯仿, 氯苯, 萘, 二甲苯, 對二氯苯, 乙苯, 1, 2 -二氯乙烷 , 四氯化碳, 苯, 反1, 2 -二氯乙烯, 順1, 2 -二氯乙烯, 三氯乙烯, 四氯乙烯, 氯乙烯, 1, 1 -二氯乙烯, 1, 1 -二氯乙烷 , 1, 1, 2 -三氯乙烷, 二氯甲烷, 氯甲烷, 甲基第三丁基醚, 1, 1, 1 -三氯乙烷, 1,2- 二氯苯, 有機氮, 懸浮固體物, 日期]
Index: []

[0 rows x 68 columns]

In [13]:
siteid

'9060212'

In [14]:
wa_df[wa_df['井號'] == int(siteid)]

Empty DataFrame
Columns: [井名, 資料來源, 日期時間, 井號, 緯度, 經度, 溫度, 電導度, 氨氮, 氫離子濃度指數, 錳, 鐵, 氯鹽, 硫酸鹽, 總有機碳, 鋅, 砷, 銅, 鎘, 鉻, 鈣, 鉛, 鈉, 鎂, 鉀, 總硬度, 總溶解固體物, 亞硝酸鹽氮, 井別, 鹼度, 水面至井口深度, 氧化還原電位, 硝酸鹽氮, 大腸桿菌群, 硫化物, 碳酸氫根, 鎳, 汞, 酚類, 氟鹽, 碳酸根, 溶氧量, 甲苯, 氯仿, 氯苯, 萘, 二甲苯, 對二氯苯, 乙苯, 1, 2 -二氯乙烷 , 四氯化碳, 苯, 反1, 2 -二氯乙烯, 順1, 2 -二氯乙烯, 三氯乙烯, 四氯乙烯, 氯乙烯, 1, 1 -二氯乙烯, 1, 1 -二氯乙烷 , 1, 1, 2 -三氯乙烷, 二氯甲烷, 氯甲烷, 甲基第三丁基醚, 1, 1, 1 -三氯乙烷, 1,2- 二氯苯, 有機氮, 懸浮固體物, 日期]
Index: []

[0 rows x 68 columns]

Okay, there is really no wate quality data of 9060212. I checked the water quality files before merging, this siteid doesn't have water measurements at all.

# Another bug

In [7]:
import visualization

# use the 2021-05-01 to 2021-05-15 as the toy data
df = pd.read_csv('data/test.csv')
select = visualization.Waterlvl()
std_name = '再生水用於工業用途水質基礎建議值一'

In [8]:
check = True
check_list=[]
out_df = pd.DataFrame()
for siteid in df['井號'].astype(str).unique():
    # collect the analytes having value in the standard (std_name) and '日期'
    cols = np.hstack([std_df.loc[~std_df[std_name].isna(), '項目'], '日期'])
    # select the data points of that siteid
    X = wa_df[wa_df['井號'] == siteid].reset_index(drop=True)
    # There might be site not having measurement at all
    if len(X) > 0:
        # select the analytes both have values in the water quality dataset
        # and in the standard.
        X = X.loc[:, X.columns.isin(cols)]
        X = X.loc[:, X.any(axis=0)].copy()
        # the last one is '日期'
        for analyte in X.columns[:-1]:
            std_value = std_df.loc[std_df['項目'] == analyte, std_name].values[0]
            # there are three different scenarios about the standard value
            if analyte == '氫離子濃度指數':
                lo_lim, up_lim = [float(_) for _ in std_value.split('-')]
                mask = (X[analyte] >= lo_lim) & (X[analyte] <= up_lim)
            elif analyte == '溶氧量':
                mask = X[analyte] > std_value
            else:
                mask = X[analyte] < std_value
            # if there more than 3 historical measurements not passing std
            # in an analyte, this siteid will be marked "no"
            if ~mask.sum() > 3: 
                check = False
    else:
        print('{} has no water quality measurement'.format(siteid))
        check = False
    check_list.append(check)
    out_df = pd.concat([out_df, df[df['井號'] == siteid]], join='outer', axis=0)

9060212 has no water quality measurement
9170231 has no water quality measurement


In [10]:
out_df.shape

(0, 7)

In [11]:
df[df['井號'] == siteid]

Empty DataFrame
Columns: [井名, 日期時間, 井號, 水面至井口深度, Lon, Lat, 月]
Index: []

In [12]:
df

井名                 日期時間        井號  水面至井口深度         Lon        Lat  \
0       國聖(1)  2021-05-01 00:00:00   7010111    3.933  120.569126  24.092658   
1       國聖(1)  2021-05-01 00:10:00   7010111    3.931  120.569126  24.092658   
2       國聖(1)  2021-05-01 00:20:00   7010111    3.932  120.569126  24.092658   
3       國聖(1)  2021-05-01 00:30:00   7010111    3.933  120.569126  24.092658   
4       國聖(1)  2021-05-01 00:40:00   7010111    3.934  120.569126  24.092658   
...       ...                  ...       ...      ...         ...        ...   
263689  東石(4)  2021-05-14 23:10:00  10090141   20.302  120.154469  23.460335   
263690  東石(4)  2021-05-14 23:20:00  10090141   20.293  120.154469  23.460335   
263691  東石(4)  2021-05-14 23:30:00  10090141   20.283  120.154469  23.460335   
263692  東石(4)  2021-05-14 23:40:00  10090141   20.274  120.154469  23.460335   
263693  東石(4)  2021-05-14 23:50:00  10090141   20.264  120.154469  23.460335   

        月  
0       5  
1       5  
2       5  
3       5  
4       5  
...    ..  
263689  5  
263690  5  
263691  5  
263692  5  
263693  5  

[263694 rows x 7 columns]

In [13]:
siteid

'10090141'

In [15]:
df['井號'].dtype

dtype('int64')

Okay, I need to transform df['井號'] to str in the begining.

# Demo

In [1]:
import visualization
import pandas as pd

# use the 2021-05-01 to 2021-05-15 as the toy data
df = pd.read_csv('data/test.csv')
select = visualization.Waterlvl()
df = select.MarkbyEP(df=df)

select = visualization.Waterquality()
out_df = select.MarkbySTD(df=df, std_name='再生水用於工業用途水質基礎建議值一')

9060212 has no water quality measurement
9170231 has no water quality measurement


In [2]:
out_df

井名        井號 水面至井口深度         Lon        Lat  月  \
2021-05-14 23:50:00  國聖(1)   7010111   3.893  120.569126  24.092658  5   
2021-05-14 23:50:00  東榮(1)  10050111  39.126  120.435512  23.557506  5   
2021-05-14 23:50:00  安和(1)  10070111  21.695  120.312655  23.514751  5   
2021-05-14 23:50:00  和豐(1)   9160212  20.494  120.231564  23.737339  5   
2021-05-13 08:20:00  芳苑(1)   7230312  14.283  120.327834  23.920473  5   
...                    ...       ...     ...         ...        ... ..   
2021-05-14 23:50:00  明德(3)   9180331  24.969  120.199259  23.652903  5   
2021-05-06 08:30:00  大溝(1)   9200211  38.773  120.210658  23.566151  5   
2021-05-14 23:50:00  文昌(4)   7060141  14.036  120.419559  24.008117  5   
2021-05-14 23:50:00  花壇(1)   7080111   2.508  120.543393  24.026659  5   
2021-05-06 12:40:00  社頭(1)   7170211    7.64  120.593717  23.900657  5   

                     wl_check  wa_check  
2021-05-14 23:50:00      True      True  
2021-05-14 23:50:00      True      True  
2021-05-14 23:50:00      True      True  
2021-05-14 23:50:00      True      True  
2021-05-13 08:20:00      True      True  
...                       ...       ...  
2021-05-14 23:50:00      True     False  
2021-05-06 08:30:00      True     False  
2021-05-14 23:50:00      True     False  
2021-05-14 23:50:00     False     False  
2021-05-06 12:40:00      True     False  

[148 rows x 8 columns]

In [3]:
out_df['井號'].duplicated().sum()

0